File: wasm/test/test_x509.mjs
/**
 * HD Wallet WASM - X.509 Tests
 */

import init, {
  Curve,
  X509Encoding,
} from '../src/index.mjs';

import { test, skip, assert, assertEqual, bytesToHex } from './test_all.mjs';

const wallet = await init();
assert(wallet.x509, 'X.509 API object must exist');
assertEqual(typeof wallet.x509.isAvailable, 'function', 'X.509 capability probe must exist');
const x509Available = wallet.x509.isAvailable();

function hexToBytes(hex) {
  const out = new Uint8Array(hex.length / 2);
  for (let i = 0; i < out.length; i++) {
    out[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
  }
  return out;
}

if (!x509Available) {
  const reason = 'OpenSSL-backed X.509 support not available in this build';
  skip('X.509: private key PEM, self-signed certificate, and encoding conversion round-trip', reason);
  skip('X.509: issuer-signed certificate and PKCS#12 import/export round-trip', reason);
} else {
  test('X.509: private key PEM, self-signed certificate, and encoding conversion round-trip', () => {
    const certPrivateKey = wallet.x509.generatePrivateKey(Curve.P256);
    const certPrivateKeyPem = wallet.x509.exportPrivateKeyPem(Curve.P256, certPrivateKey);
    const importedPrivateKey = wallet.x509.importPrivateKeyPem(Curve.P256, certPrivateKeyPem);

    assert(certPrivateKeyPem.includes('BEGIN PRIVATE KEY'), 'Expected PKCS#8 PEM output');
    assertEqual(bytesToHex(importedPrivateKey), bytesToHex(certPrivateKey), 'Private key PEM round-trip should match');

    const certificatePem = wallet.x509.createSelfSignedCertificate(
      {
        subjectDn: 'CN=wallet.example,O=HD Wallet,C=US',
        serialHex: '0102A0',
        notBeforeUnix: 1704067200,
        notAfterUnix: 1735689600,
        dnsNames: ['wallet.example', 'api.wallet.example'],
        keyUsage: ['digitalSignature', 'keyEncipherment'],
        extendedKeyUsage: ['serverAuth', 'clientAuth'],
        walletAttestation: {
          curve: Curve.SECP256K1,
          privateKey: hexToBytes('0000000000000000000000000000000000000000000000000000000000000001'),
          keyLabel: 'btc-main-0',
          commentPrefix: 'HDWALLET-X509'
        }
      },
      Curve.P256,
      certPrivateKey,
      X509Encoding.PEM
    );

    assert(typeof certificatePem === 'string', 'PEM certificate should be returned as string');
    assert(certificatePem.includes('BEGIN CERTIFICATE'), 'Expected PEM certificate output');

    const parsed = wallet.x509.parseCertificate(certificatePem, X509Encoding.PEM);
    assert(parsed.subjectDn.includes('CN=wallet.example'), 'Expected parsed subject DN');
    assertEqual(parsed.walletAttestationValid, true, 'Wallet attestation should validate');
    assertEqual(wallet.x509.verifyWalletAttestation(certificatePem, X509Encoding.PEM), true);

    const certificateDer = wallet.x509.convertCertificate(
      certificatePem,
      X509Encoding.PEM,
      X509Encoding.DER
    );
    assert(certificateDer instanceof Uint8Array, 'DER certificate should be returned as bytes');
    assert(certificateDer.length > 0, 'DER certificate should not be empty');

    const roundTripPem = wallet.x509.convertCertificate(
      certificateDer,
      X509Encoding.DER,
      X509Encoding.PEM
    );
    assert(typeof roundTripPem === 'string' && roundTripPem.includes('BEGIN CERTIFICATE'));
  });

  test('X.509: issuer-signed certificate and PKCS#12 import/export round-trip', () => {
    for (let i = 0; i < 5; i++) {
      const rootPrivateKey = wallet.x509.generatePrivateKey(Curve.P384);
      const leafPrivateKey = wallet.x509.generatePrivateKey(Curve.P384);

      const rootCertificatePem = wallet.x509.createSelfSignedCertificate(
        {
          subjectDn: 'CN=Wallet Root CA,O=HD Wallet,C=US',
          serialHex: '1001',
          notBeforeUnix: 1704067200,
          notAfterUnix: 1767225600,
          isCa: true,
          pathLen: 0,
          keyUsage: ['keyCertSign', 'cRLSign'],
          friendlyName: 'wallet-root'
        },
        Curve.P384,
        rootPrivateKey,
        X509Encoding.PEM
      );

      const leafCertificatePem = wallet.x509.issueCertificate(
        {
          subjectDn: 'CN=leaf.wallet.example,O=HD Wallet,C=US',
          serialHex: '1002',
          notBeforeUnix: 1704067200,
          notAfterUnix: 1735689600,
          dnsNames: ['leaf.wallet.example'],
          keyUsage: ['digitalSignature', 'keyEncipherment'],
          extendedKeyUsage: ['serverAuth'],
          friendlyName: 'wallet-leaf'
        },
        Curve.P384,
        rootPrivateKey,
        rootCertificatePem,
        X509Encoding.PEM,
        Curve.P384,
        leafPrivateKey,
        X509Encoding.PEM
      );

      const pkcs12 = wallet.x509.exportPkcs12(
        leafCertificatePem,
        X509Encoding.PEM,
        Curve.P384,
        leafPrivateKey,
        'changeit',
        'wallet-leaf',
        rootCertificatePem
      );

      assert(pkcs12 instanceof Uint8Array, `PKCS#12 export should return bytes (iteration ${i + 1})`);
      assert(pkcs12.length > 0, `PKCS#12 export should not be empty (iteration ${i + 1})`);

      const imported = wallet.x509.importPkcs12(pkcs12, 'changeit');
      assert(imported.certificatePem.includes('BEGIN CERTIFICATE'), `Imported PKCS#12 should include certificate PEM (iteration ${i + 1})`);
      assert(imported.privateKeyPem.includes('BEGIN PRIVATE KEY'), `Imported PKCS#12 should include private key PEM (iteration ${i + 1})`);
      assert(imported.chainPem.includes('BEGIN CERTIFICATE'), `Imported PKCS#12 should include chain PEM (iteration ${i + 1})`);
    }
  });
}

File: test/test_x509.cpp
/**
 * @file test_x509.cpp
 * @brief X.509 certificate tests
 */

#include "test_framework.h"

#include "hd_wallet/x509.h"

#include <ctime>

using namespace hd_wallet;

namespace {

ByteVector makeWalletKey() {
  ByteVector key(32);
  for (size_t i = 0; i < key.size(); ++i) {
    key[i] = static_cast<uint8_t>(i + 1);
  }
  return key;
}

x509::CertificateOptions makeBaseOptions(const std::string& subject_dn) {
  x509::CertificateOptions options;
  options.subject_dn = subject_dn;
  options.serial_hex = "01";
  const auto now = static_cast<int64_t>(std::time(nullptr));
  options.not_before_unix = now - 60;
  options.not_after_unix = now + 86400;
  return options;
}

} // namespace

TEST_CASE(X509, PrivateKeyPemRoundTrip_P256) {
  auto private_key = x509::generatePrivateKey(Curve::P256);
  ASSERT_OK(private_key);
  ASSERT_EQ(32u, private_key.value.size());

  auto pem = x509::exportPrivateKeyPem(Curve::P256, private_key.value);
  ASSERT_OK(pem);
  ASSERT_NE(std::string::npos, pem.value.find("BEGIN PRIVATE KEY"));

  auto imported = x509::importPrivateKeyPem(Curve::P256, pem.value);
  ASSERT_OK(imported);
  ASSERT_EQ(private_key.value.size(), imported.value.size());
  ASSERT_BYTES_EQ(private_key.value.data(), imported.value.data(), private_key.value.size());
}

TEST_CASE(X509, SelfSignedCertificate_WalletAttestation) {
  auto cert_key = x509::generatePrivateKey(Curve::P256);
  ASSERT_OK(cert_key);

  auto options = makeBaseOptions("CN=wallet.example.com,O=Digital Arsenal,C=US");
  options.dns_names = {"wallet.example.com", "www.wallet.example.com"};
  options.key_usage = {"digitalSignature", "keyEncipherment"};
  options.extended_key_usage = {"serverAuth"};
  options.include_wallet_attestation = true;
  options.wallet_attestation.curve = Curve::SECP256K1;
  options.wallet_attestation.private_key = makeWalletKey();
  options.wallet_attestation.key_label = "btc-root";

  auto certificate = x509::createSelfSignedCertificate(
      options,
      Curve::P256,
      cert_key.value,
      x509::Encoding::PEM);
  ASSERT_OK(certificate);
  const std::string certificate_pem(certificate.value.begin(), certificate.value.end());
  ASSERT_NE(std::string::npos, certificate_pem.find("BEGIN CERTIFICATE"));

  auto verified = x509::verifyWalletAttestation(certificate.value, x509::Encoding::PEM);
  ASSERT_OK(verified);
  ASSERT_TRUE(verified.value);

  auto parsed = x509::parseCertificateJson(certificate.value, x509::Encoding::PEM);
  ASSERT_OK(parsed);
  ASSERT_NE(std::string::npos, parsed.value.find("wallet.example.com"));
  ASSERT_NE(std::string::npos, parsed.value.find("\"wallet_attestation_valid\":true"));
  ASSERT_NE(std::string::npos, parsed.value.find("btc-root"));
}

TEST_CASE(X509, IssuerSignedCertificate_P384_AndPkcs12RoundTrip) {
  auto issuer_key = x509::generatePrivateKey(Curve::P384);
  ASSERT_OK(issuer_key);
  auto subject_key = x509::generatePrivateKey(Curve::P384);
  ASSERT_OK(subject_key);

  auto issuer_options = makeBaseOptions("CN=Test Root CA,O=Digital Arsenal,C=US");
  issuer_options.serial_hex = "CA01";
  issuer_options.is_ca = true;
  issuer_options.path_len = 0;
  issuer_options.key_usage = {"keyCertSign", "cRLSign"};

  auto issuer_certificate = x509::createSelfSignedCertificate(
      issuer_options,
      Curve::P384,
      issuer_key.value,
      x509::Encoding::PEM);
  ASSERT_OK(issuer_certificate);

  auto subject_options = makeBaseOptions("CN=leaf.example.com,O=Digital Arsenal,C=US");
  subject_options.serial_hex = "1001";
  subject_options.dns_names = {"leaf.example.com"};
  subject_options.key_usage = {"digitalSignature", "keyEncipherment"};
  subject_options.extended_key_usage = {"serverAuth"};

  auto leaf_certificate = x509::issueCertificate(
      subject_options,
      Curve::P384,
      issuer_key.value,
      issuer_certificate.value,
      x509::Encoding::PEM,
      Curve::P384,
      subject_key.value,
      x509::Encoding::PEM);
  ASSERT_OK(leaf_certificate);

  auto der_leaf = x509::convertCertificateEncoding(
      leaf_certificate.value,
      x509::Encoding::PEM,
      x509::Encoding::DER);
  ASSERT_OK(der_leaf);
  ASSERT_TRUE(!der_leaf.value.empty());

  auto parsed = x509::parseCertificateJson(leaf_certificate.value, x509::Encoding::PEM);
  ASSERT_OK(parsed);
  ASSERT_NE(std::string::npos, parsed.value.find("leaf.example.com"));
  ASSERT_NE(std::string::npos, parsed.value.find("Test Root CA"));

  auto pkcs12 = x509::exportPkcs12(
      leaf_certificate.value,
      x509::Encoding::PEM,
      Curve::P384,
      subject_key.value,
      "secret",
      "leaf-cert");
  ASSERT_OK(pkcs12);
  ASSERT_TRUE(!pkcs12.value.empty());

  auto imported = x509::importPkcs12Json(pkcs12.value, "secret");
  ASSERT_OK(imported);
  ASSERT_NE(std::string::npos, imported.value.find("BEGIN CERTIFICATE"));
  ASSERT_NE(std::string::npos, imported.value.find("BEGIN PRIVATE KEY"));
}
